python/cugraph/cugraph/dask/structure/mg_property_graph.py

```diff
@@ -448,27 +425,38 @@ def add_vertex_data(
         # Save the original dtypes for each new column so they can be restored
         # prior to constructing subgraphs (since column dtypes may get altered
         # during merge to accommodate NaN values).
-        new_col_info = self.__get_new_column_dtypes(
-            tmp_df, self.__vertex_prop_dataframe
-        )
+        if is_first_data:
+            new_col_info = tmp_df.dtypes.items()
+        else:
+            new_col_info = self.__get_new_column_dtypes(
+                tmp_df, self.__vertex_prop_dataframe
+            )
         self.__vertex_prop_dtypes.update(new_col_info)
 
-        # Join on shared columns and the indices
+        # TODO: allow tmp_df to come in with vertex id already as index
         tmp_df = tmp_df.persist().set_index(self.vertex_col_name).persist()
-        cols = self.__vertex_prop_dataframe.columns.intersection(
-            tmp_df.columns
-        ).to_list()
-        cols.append(self.vertex_col_name)
-        # FIXME: workaround for: https://github.com/rapidsai/cudf/issues/11550
-        self.__vertex_prop_dataframe = (
-            self.__vertex_prop_dataframe.reset_index()
-            .merge(tmp_df.reset_index(), on=cols, how="outer")
-            .persist()
-            .set_index(self.vertex_col_name)
-            .persist()
-        )
-        # self.__vertex_prop_dataframe = \
-        #     self.__vertex_prop_dataframe.merge(tmp_df, on=cols, how="outer")
+        self.__update_dataframe_dtypes(tmp_df, self.__vertex_prop_dtypes)
+
+        if is_first_data:
+            self.__vertex_prop_dataframe = tmp_df
+        else:
+            # Join on vertex ids (the index)
+            # TODO: can we automagically determine when we to use concat?
+            df = self.__vertex_prop_dataframe.join(
+                tmp_df, how="outer", rsuffix="_NEW_", npartitions=self.__num_workers
+            ).persist()
+            cols = self.__vertex_prop_dataframe.columns.intersection(
+                tmp_df.columns
+            ).to_list()
+            rename_cols = {f"{col}_NEW_": col for col in cols}
+            new_cols = list(rename_cols)
+            sub_df = df[new_cols].rename(columns=rename_cols)
+            # This only adds data--it doesn't replace existing data
+            df = df.drop(columns=new_cols).fillna(sub_df).persist()
+            if df.npartitions > 2 * self.__num_workers:
+                # TODO: better understand behavior of npartitions argument in join
+                df = df.repartition(npartitions=self.__num_workers).persist()
```

Comment:
With `dask_cudf` merges/operations,  I think we should actually decrease number of partitions to  less than `_num_workers` unless we actually need to because it decrease parallelization and worker starving can become a problem.  

Though the shuffle cost of the merge is nlogn but if your input partitions are less than availble workers , worker starving can become a problem. 

Maybe lets pick `2*_num_workers` or something ?